### PR TITLE
add api required by wechall 

### PR DESCRIPTION
**Pull request by Zeeshan12340**
This is focused on #184  . The first two steps are implemented, although the scoring output only satisfies the minimal requirements of Step 2. Global rank and scoring is somewhat of a mess with no proper weighting system of challenges resulting in `babyarch` and `babysuid` having the same single point for solved/unsolved which is not really fair.

But, this does complete the minimum requirements. Sample output of the additions is below:
- /validate endpoint
```bash
zeeshan@hydra:/opt/dojo$ curl 'http://localhost/pwncollege_api/v1/validate?username=user&email=user@test.com'
1
zeeshan@hydra:/opt/dojo$ curl 'http://localhost/pwncollege_api/v1/validate?username=user&email=user@doesnotexist.com'
0
```

- /score endpoint
```bash
zeeshan@hydra:/opt/dojo$ curl 'http://localhost/pwncollege_api/v1/score?username=user'
"15:104"
zeeshan@hydra:/opt/dojo$ curl 'http://localhost/pwncollege_api/v1/score?username=admin'
"0:104"
zeeshan@hydra:/opt/dojo$ curl 'http://localhost/pwncollege_api/v1/score?username=adm'
{
    "error": "user does not exist"
}
```

### CODE REVIEW

File: dojo_plugin/api/v1/score.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    for dojo in official_dojos:
+        for challenge in dojo.challenges:
+            official_challenge_ids.append(challenge.challenge_id)
+
+        max_score += len(dojo.challenges)
+
+    for solve in user.solves:
+        if solve.challenge_id in official_challenge_ids:
+            user_score += 1
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/score.py:32]: Three lines:

```py
official_challenge_ids = {challenge.challenge_id for dojo in Dojos.query.filter_by(official=1).all() for challenge in dojo.challenges}
user_score = sum(1 for solve in user.solves if solve.challenge_id in official_challenge_ids)
max_score = len(official_challenge_ids)
```

File: dojo_plugin/api/v1/score.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /score?username=user
+    This endpoint is publicly available with no auth needed.
+    Returns formatted data regarding a user's score.
+    """
+    def get(self):
+        username = request.args.get("username")
+        if not username:
+            return {"error": "`username` parameter is required"}, 400
+
+        return get_user_score(username=username)
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/score.py:48]: Lets throw the code in here, boo to single-use helper functions.

File: dojo_plugin/api/v1/validate.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from flask import request
+from flask_restx import Namespace, Resource
+from CTFd.cache import cache
+from CTFd.models import Users
+
+validate_namespace = Namespace(
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/validate.py:6]: Lets throw this entire endpoint into score.py / score_endpoint.

File: dojo_plugin/api/v1/validate.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    This endpoint is publicly available, no auth needed.
+    Returns 1 or 0 depending on if username/email exists or not.
+    """
+    def get(self):
+        username = request.args.get("username")
+        email = request.args.get("email")
+        if not username or not email:
+            return {"error": "`username` and `email` parameters are required"}, 400
+
+        return get_validate_user(username=username, email=email)
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/validate.py:31]: Again no single-use helpers.

File: dojo_plugin/api/v1/validate.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    "validate", description="Endpoint to validate if user exists"
+)
+
+def get_validate_user(username, email):
+    user = Users.query.filter_by(name=username, email=email).first()
+
+    if user:
+        return 1
+    else:
+        return 0
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/validate.py:16]:
```py
int(bool(Users.query.filter_by(name=username, email=email).first()))
```

File: dojo_plugin/api/v1/validate.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+@validate_namespace.route("")
+class ValidateUser(Resource):
+    """
+    /validate?username=user&email=user@test.com
+    This endpoint is publicly available, no auth needed.
+    Returns 1 or 0 depending on if username/email exists or not.
+    """
+    def get(self):
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/validate.py:25]: Can we also rate limit this endpoint? I'm concerned about someone being dumb and trying to bruteforce discover someone's email.


```py
from CTFd.utils.decorators import ratelimit
@ratelimit(method="GET", limit=10, interval=60)

File: dojo_plugin/api/v1/score.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+from ...models import Dojos
+
+score_namespace = Namespace("score")
+
+def get_user_score(username):
+    # Alternative: hardcode max count and challenge id's for speed
+    # POSSIBLE BUG HERE - topics and courses dojos might result in doubling
+    # but since solves carry over, it should not be an issue
+    # Maybe have a proper scoring system where babyarch has more weight than babysuid
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/score.py:14]: I agree with this concern, but I also agree that we can just go ahead and keep it simple for now.

File: dojo_plugin/api/v1/score.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+from CTFd.cache import cache
+from CTFd.models import Users
+
+from ...models import Dojos
+
+score_namespace = Namespace("score")
+
+def get_user_score(username):
+    # Alternative: hardcode max count and challenge id's for speed
+    # POSSIBLE BUG HERE - topics and courses dojos might result in doubling
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/score.py:12]: Changing challenge_ids to a set as per my change recommendation makes it clear this won't happen.

File: dojo_plugin/api/v1/score.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+from flask_restx import Namespace, Resource
+from CTFd.cache import cache
+from CTFd.models import Users
+
+from ...models import Dojos
+
+score_namespace = Namespace("score")
+
+def get_user_score(username):
+    # Alternative: hardcode max count and challenge id's for speed
```
Comment (ConnorNelson) [inline dojo_plugin/api/v1/score.py:11]: It's fast enough! :)

**Review by ConnorNelson — CHANGES_REQUESTED** [review]
Thanks for the PR! Please take a look at the requested changes (especially the ratelimit), and I'll go ahead and merge :)

**Comment by Zeeshan12340** [comment]
So, I think I got all of them, let me know if I missed anything.

**Comment by Zeeshan12340** [comment]
The email is not even required for checking if the account exists and I added that just so that it matches up with the Step 1 requirement, maybe that can just be simply removed if the wechall people do not require it.

**Comment by ConnorNelson** [comment]
Look great! Thanks :)

**Pull request merged**